### ISSUE TITLE
libgit2 fails to read config files with UTF8-BOM

### ISSUE THREAD

**Issue by csware** [libgit2/libgit2#4516]
### Reproduction steps

1. Create a .gitconfig with a UTF-8 BOM
2. Open an arbitrary repository (using `git_repository_open`)

### Expected behavior

Config file is parsed correctly (as Git for Windows does or older libgit2 versions such as e0568621535869451c91e915d28a69ac7c84b8b7).

### Actual behavior

libgit2 fails with the error: `failed to parse config file: Invalid configuration key (in C:/Users/Username/.gitconfig:1, column 0)`.

In debug mode I get the following assertion:

> ---------------------------
> Microsoft Visual C++ Runtime Library
> ---------------------------
> Debug Assertion Failed!
> 
> Program: D:\TortoiseGit\src\..\bin\Debug64\bin\TortoiseGitProc.exe
> File: minkernel\crts\ucrt\src\appcrt\convert\isctype.cpp
> Line: 42
> 
> Expression: c >= -1 && c <= 255

### Version of libgit2 (release number or SHA1)

f7225946edeae35f48e3d402e2e0c94ea07f9666

### Operating system(s) tested

Windows 7 x64

### Additional notes ###

This affects all files which are parsed as config files, such as `.gitmodules`...

Having UTF-8 encoded umlauts is not a problem.

Maybe this is related to compiler: I'm using MS VisualStudio 2017 15.5.5

#### Stacktrace: ####

```
ucrtbased.dll!_chvalidator_l(__crt_locale_pointers * locale=0x000000000021b150, int c=-17, int mask=263)
ucrtbased.dll!_isalnum_l(int c=-17, __crt_locale_pointers * locale=0x0000000000000000)
ucrtbased.dll!isalnum(int c=-17)
libgit2_tgit.dll!is_namechar(char c='ï')
libgit2_tgit.dll!parse_name(char * * name=0x000000000021b368, const char * * value=0x000000000021b268, git_config_parser * reader=0x000000000021b460, const char * line=0x0000000003e6c820)
libgit2_tgit.dll!parse_variable(git_config_parser * reader=0x000000000021b460, char * * var_name=0x000000000021b368, char * * var_value=0x000000000021b388)
libgit2_tgit.dll!git_config_parse(git_config_parser * parser=0x000000000021b460, int(*)(git_config_parser *, const char *, const char *, unsigned __int64, void *) on_section=0x0000000000000000, int(*)(git_config_parser *, const char *, char *, char *, const char *, unsigned __int64, void *) on_variable=0x000007fedcb74830, int(*)(git_config_parser *, const char *, unsigned __int64, void *) on_comment=0x0000000000000000, int(*)(git_config_parser *, const char *, void *) on_eof=0x0000000000000000, void * data=0x000000000021b418)
libgit2_tgit.dll!config_read(kh_str_s * values=0x0000000003e515b0, const git_repository * repo=0x0000000003e5b970, config_file * file=0x0000000003e3a918, git_config_level_t level=GIT_CONFIG_LEVEL_GLOBAL, int depth=0)
libgit2_tgit.dll!config_open(git_config_backend * cfg=0x0000000003e3a760, git_config_level_t level=GIT_CONFIG_LEVEL_GLOBAL, const git_repository * repo=0x0000000003e5b970)
libgit2_tgit.dll!git_config_add_backend(git_config * cfg=0x0000000000392320, git_config_backend * file=0x0000000003e3a760, git_config_level_t level=GIT_CONFIG_LEVEL_GLOBAL, const git_repository * repo=0x0000000003e5b970, int force=0)
libgit2_tgit.dll!git_config_add_file_ondisk(git_config * cfg=0x0000000000392320, const char * path=0x0000000003e514d0, git_config_level_t level=GIT_CONFIG_LEVEL_GLOBAL, const git_repository * repo=0x0000000003e5b970, int force=0)
libgit2_tgit.dll!load_config(git_config * * out=0x000000000021b808, git_repository * repo=0x0000000003e5b970, const char * global_config_path=0x0000000003e514d0, const char * xdg_config_path=0x0000000000000000, const char * system_config_path=0x0000000000391fa0, const char * programdata_path=0x0000000003e51540)
libgit2_tgit.dll!git_repository_config__weakptr(git_config * * out=0x000000000021b878, git_repository * repo=0x0000000003e5b970)
libgit2_tgit.dll!git_repository_config_snapshot(git_config * * out=0x000000000021b9d8, git_repository * repo=0x0000000003e5b970)
libgit2_tgit.dll!git_repository_open_ext(git_repository * * repo_ptr=0x000000000021bb58, const char * start_path=0x00000000003c9be8, unsigned int flags=1, const char * ceiling_dirs=0x0000000000000000)
libgit2_tgit.dll!git_repository_open(git_repository * * repo_out=0x000000000021bb58, const char * path=0x00000000003c9be8)
```

Cross-ref: https://gitlab.com/tortoisegit/tortoisegit/issues/3155

**Comment by ethomson** [comment]
Ugh.  UTF-8 BOMs, huh?

You haven't mentioned it explicitly, but I assume that Git does not fail to open these files?

**Comment by csware** [comment]
> You haven't mentioned it explicitly, but I assume that Git does not fail to open these files?

Git for WIndows does (as mentioned by expected bahavior).

**Comment by ethomson** [comment]
Yes, thanks, I see that now.  I suppose this is due to PowerShell's monstrous inabillity to cope with text files correctly.  I suppose we'll have to fix this.

**Comment by csware** [comment]
This is a regression e0568621535869451c91e915d28a69ac7c84b8b7 worked just fine.

**Comment by csware** [comment]
The regression seems to be introduced between 88450c1c6043b34c4433c475dcfd4928439a4810 (working) and 15f8d9a232097c63c87d8bb2b959e72bb7c1c798 (not working any more).

**Comment by pks-t** [comment]
Thanks for the report, @csware. The issue should be fixed by #4521, as well. Please test

**Comment by csware** [comment]
Fixed by PR #4521.
